pkg/applicationserver/device_fetcher.go

```diff
@@ -69,22 +73,129 @@ type cachedEndDeviceFetcher struct {
+func NewSingleFlightEndDeviceFetcher(fetcher EndDeviceFetcher) EndDeviceFetcher {
+	return &singleFlightEndDeviceFetcher{fetcher: fetcher}
+}
+
+// Get implements the EndDeviceFetcher interface.
+func (f *singleFlightEndDeviceFetcher) Get(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, fieldMaskPaths ...string) (*ttnpb.EndDevice, error) {
+	key := endDeviceKey(ctx, ids, fieldMaskPaths...)
+	dev, err, _ := f.singleflight.Do(key, func() (interface{}, error) {
+		return f.fetcher.Get(ctx, ids, fieldMaskPaths...)
+	})
 	if err != nil {
-		dev, err := f.fetcher.Get(ctx, ids, fieldMaskPaths...)
-		if err == nil {
-			f.cache.Set(key, dev)
-		}
-		return dev, err
+		return nil, err
 	}
-	return dev.(*ttnpb.EndDevice), err
+	return dev.(*ttnpb.EndDevice), nil
+}
+
+type timeoutEndDeviceFetcher struct {
+	fetcher EndDeviceFetcher
+	timeout time.Duration
+}
+
+// NewTimeoutEndDeviceFetcher wraps an EndDeviceFetcher and limits the lifetime of the context used to retrieve the end device.
+func NewTimeoutEndDeviceFetcher(fetcher EndDeviceFetcher, timeout time.Duration) EndDeviceFetcher {
+	return &timeoutEndDeviceFetcher{fetcher, timeout}
+}
+
+// Get implements the EndDeviceFetcher interface.
+func (f *timeoutEndDeviceFetcher) Get(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, fieldMaskPaths ...string) (*ttnpb.EndDevice, error) {
+	ctx, cancel := context.WithTimeout(ctx, f.timeout)
+	defer cancel()
+	return f.fetcher.Get(ctx, ids, fieldMaskPaths...)
+}
+
+type circuitBreakerEndDeviceFetcher struct {
+	fetcher EndDeviceFetcher
+
+	threshold uint64
+	timeout   time.Duration
+
+	mu                sync.RWMutex
+	failures          uint64
+	lastFailedAttempt time.Time
+}
+
+// NewCircuitBreakerEndDeviceFetcher wraps an end device fetcher with a circuit breaking mechanism.
+// The circuit breaker opens when the number of failure attempts is higher than the threshold,
+// and closes after the provided timeout.
+func NewCircuitBreakerEndDeviceFetcher(fetcher EndDeviceFetcher, threshold uint64, timeout time.Duration) EndDeviceFetcher {
+	return &circuitBreakerEndDeviceFetcher{
+		fetcher:   fetcher,
+		threshold: threshold,
+		timeout:   timeout,
+	}
+}
+
+var errCircuitBreakerOpen = errors.DefineUnavailable("circuit_breaker_open", "circuit breaker open")
+
+func (f *circuitBreakerEndDeviceFetcher) circuitOpen() error {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	// If the number of failures is still under the threshold, consider the circuit as being closed.
+	if f.failures < f.threshold {
+		return nil
+	}
+	// If the attempt timeout expired, consider the circuit as being closed.
+	if time.Now().Sub(f.lastFailedAttempt) > f.timeout {
+		return nil
+	}
+	// At this point we have a number of failures that is above the threshold, and any attempts are recent.
+	// The circuit breaker is open.
+	return errCircuitBreakerOpen.New()
+}
+
+func (f *circuitBreakerEndDeviceFetcher) observeError(err error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if err != nil {
```

Comment:
I think we should consider the error type here. It's relatively easy to trigger NotFound errors by deleting an end device from the IS without deleting it from the AS. If you'd coordinate this across multiple devices you can relatively easily trigger the circuit breaker for other users in the network.